zcash_client_backend/src/zip321.rs

```diff
@@ -578,21 +576,30 @@ pub mod parse {
         alphanum_or("+-".to_string())(input)
     }
 
-    pub fn indexed_name<'a>(input: &'a str) -> IResult<&'a str, (&'a str, Option<&'a str>)> {
+    pub fn indexed_name(input: &str) -> IResult<&str, (&str, Option<&str>)> {
         let paramname = recognize(tuple((alpha1, namechars)));
 
         tuple((
             paramname,
             opt(preceded(
                 char('.'),
-                recognize(tuple((one_of("123456789"), digit0))),
+                recognize(tuple((
+                    one_of("123456789"),
+                    map_opt(digit0, |s: &str| if s.len() > 3 { None } else { Some(s) }),
```

Comment:
Isn't [take_while_m_n!](https://docs.rs/nom/5.1.2/nom/macro.take_while_m_n.html) a clearer way to do this?

Reply:
That's a byte-level combinator, not a character-level one. I actually tried using it here, but the cascade of replacing `&str` with `&[u8]` actually made for a lot more encoding and decoding, so I decided this was cleaner.